File: desktop/src/bridge.ts
import { invoke } from "@tauri-apps/api/core";

type BridgeEnvelope<T> = {
  ok: boolean;
  data?: T;
  error?: string;
};

type BridgePayload = {
  project_id?: string;
  run_id?: string;
  job_id?: string;
  check_id?: string;
  dry_run?: boolean;
  global_enabled?: boolean;
  project_enabled?: boolean;
  backend?: string;
  search_top_k?: number;
  search_threshold?: number;
  limit?: number;
  provider_id?: string;
  env_values?: Record<string, string>;
  preferred?: boolean;
  prompt?: string;
  project_name?: string;
  project_path?: string;
  primary_agent?: string;
};

const mockBridgeData = {
  health: {
    version: "0.5.0",
    platform: "desktop-dev",
    workspace_count: 0,
    available_agents: 0,
    available_providers: 0,
  },
  projects: [],
  providers: [],
  setup_status: [
    {
      check_id: "python",
      label: "Python",
      detected: true,
      required: true,
      version: "Python 3.12.0",
      reason: "",
      install_hint: "Install Python 3.11+ and keep it on PATH for the desktop sidecar.",
      install_commands: ["winget install Python.Python.3.11"],
    },
    {
      check_id: "cargo",
      label: "Cargo",
      detected: false,
      required: false,
      version: "",
      reason: "Cargo was not found on PATH.",
      install_hint: "Install Cargo via rustup to build native desktop packages.",
      install_commands: ["winget install Rustlang.Rustup"],
    },
  ],
  runs: [],
  create_project: {
    project_id: "mock-project",
    name: "Mock Project",
    project_path: "C:/mock/project",
    primary_agent: "codex",
    last_active: "",
    session_id: null,
    latest_run_id: "",
    latest_run_status: "",
    latest_run_phase: "",
    latest_run_completion: "",
    needs_follow_up: false,
    attention_tags: [],
    priority_rank: 0,
    triage_summary: "",
    triage_evidence: [],
  },
  memory_status: {
    project_id: "mock-project",
    project_path: "C:/mock/project",
    enabled: true,
    global_enabled: true,
    project_enabled: true,
    backend: "file",
    mem0_available: false,
    current_backend: "file",
    total_memories: 2,
    legacy_entry_count: 1,
    migration_ready: false,
    migration_issues: ["mem0 unavailable in mock bridge"],
    memory_sources: { hypotheses: 0, learnings: 0, project_memory: 1 },
    init_error: "",
  },
  memory_records: [
    {
      memory_id: "mem-001",
      content: "Workflow completion for mock run",
      memory_type: "workflow_completion",
      phase: "complete",
      run_id: "mock-sprint",
      timestamp: "2026-03-22T00:00:00Z",
    },
  ],
  memory_migrate: {
    project_id: "mock-project",
    dry_run: true,
    success: true,
    total_migrated: 1,
    source_counts: { hypotheses: 0, learnings: 0, project_memory: 1 },
    errors: [],
  },
  memory_settings: {
    project_id: "mock-project",
    global_enabled: true,
    project_enabled: true,
    effective_enabled: true,
    backend: "file",
    search_top_k: 5,
    search_threshold: 0.7,
    embedding_model: "text-embedding-3-small",
    embedding_dims: 1536,
  },
  memory_update_settings: {
    project_id: "mock-project",
    global_enabled: true,
    project_enabled: true,
    effective_enabled: true,
    backend: "mem0_local",
    search_top_k: 8,
    search_threshold: 0.6,
    embedding_model: "text-embedding-3-small",
    embedding_dims: 1536,
  },
  update_provider: {
    provider_id: "mock",
    label: "Mock",
    detected: false,
    auth_ok: false,
    binary_path: null,
    version: null,
    reason: "",
    install_commands: [],
    required_env_vars: [],
    configured_env_vars: [],
    missing_env_vars: [],
    stored_env_vars: {},
    preferred: false,
  },
  install_provider: {
    provider_id: "mock",
    command: "npm install -g mock",
    success: true,
    summary: "Mock install completed.",
    output: "installed",
    next_steps: ["mock auth login"],
  },
  install_setup_dependency: {
    check_id: "cargo",
    command: "winget install Rustlang.Rustup",
    success: true,
    summary: "Cargo install completed.",
    output: "installed",
    next_steps: [],
  },
  start_run: {
    project_id: "mock-project",
    sprint_id: "mock-sprint",
    success: true,
    state: "completed",
    summary: "Mock desktop run completed.",
    next_steps: ["Inspect workflow artifacts"],
  },
  start_run_async: {
    job_id: "job-mock-001",
    project_id: "mock-project",
    prompt: "Mock prompt",
    status: "queued",
    created_at: "2026-03-22T00:00:00Z",
    updated_at: "2026-03-22T00:00:00Z",
    sprint_id: "",
    summary: "",
    state: "",
    next_steps: [],
    error: "",
  },
  run_job_status: {
    job_id: "job-mock-001",
    project_id: "mock-project",
    prompt: "Mock prompt",
    status: "completed",
    created_at: "2026-03-22T00:00:00Z",
    updated_at: "2026-03-22T00:00:10Z",
    sprint_id: "mock-sprint",
    summary: "Mock desktop run completed.",
    state: "completed",
    next_steps: ["Inspect workflow artifacts"],
    error: "",
  },
};

export async function invokeAppRuntime<T>(action: string, payload: BridgePayload = {}): Promise<T> {
  const isTauri = typeof window !== "undefined" && "__TAURI_INTERNALS__" in window;

  if (!isTauri) {
    return mockBridgeData[action as keyof typeof mockBridgeData] as T;
  }

  const raw = await invoke<string>("app_runtime", {
    action,
    projectId: payload.project_id,
    runId: payload.run_id,
    jobId: payload.job_id,
    checkId: payload.check_id,
    dryRun: payload.dry_run,
    globalEnabled: payload.global_enabled,
    projectEnabled: payload.project_enabled,
    backend: payload.backend,
    searchTopK: payload.search_top_k,
    searchThreshold: payload.search_threshold,
    limit: payload.limit,
    providerId: payload.provider_id,
    envValues: payload.env_values,
    preferred: payload.preferred,
    prompt: payload.prompt,
    projectName: payload.project_name,
    projectPath: payload.project_path,
    primaryAgent: payload.primary_agent,
  });
  const envelope = JSON.parse(raw) as BridgeEnvelope<T>;
  if (!envelope.ok || envelope.data === undefined) {
    throw new Error(envelope.error || `Bridge call failed for action: ${action}`);
  }
  return envelope.data;
}

File: src/aop/app_runtime/bridge.py
"""Simple JSON bridge for desktop app experiments."""

from __future__ import annotations

from typing import Any, Dict

from .service import DesktopAppService


class DesktopAppBridge:
    """Dispatch simple desktop bridge actions."""

    def __init__(self, service: DesktopAppService | None = None) -> None:
        self.service = service or DesktopAppService()

    def dispatch(self, action: str, payload: Dict[str, Any] | None = None) -> Dict[str, Any]:
        """Execute one bridge action and return JSON-serializable data."""
        data = payload or {}
        if action == "health":
            return {"ok": True, "data": self.service.get_app_health().to_dict()}
        if action == "projects":
            return {
                "ok": True,
                "data": [project.to_dict() for project in self.service.list_projects()],
            }
        if action == "providers":
            return {
                "ok": True,
                "data": [provider.to_dict() for provider in self.service.get_provider_status()],
            }
        if action == "create_project":
            project_name = str(data.get("project_name", "")).strip()
            project_path = str(data.get("project_path", "")).strip()
            primary_agent = str(data.get("primary_agent", "codex")).strip() or "codex"
            try:
                project = self.service.create_project(
                    name=project_name,
                    project_path=project_path,
                    primary_agent=primary_agent,
                )
            except ValueError as error:
                return {"ok": False, "error": str(error)}
            return {"ok": True, "data": project.to_dict()}
        if action == "settings":
            return {"ok": True, "data": self.service.get_settings()}
        if action == "setup_status":
            return {
                "ok": True,
                "data": [check.to_dict() for check in self.service.get_setup_status()],
            }
        if action == "runs":
            project_id = str(data.get("project_id", "")).strip()
            runs = self.service.list_runs(project_id=project_id, limit=int(data.get("limit", 12)))
            return {
                "ok": True,
                "data": [run.__dict__ for run in runs],
            }
        if action == "run_detail":
            project_id = str(data.get("project_id", "")).strip()
            run_id = str(data.get("run_id", "")).strip()
            detail = self.service.get_run_detail(project_id=project_id, run_id=run_id)
            if detail is None:
                return {"ok": False, "error": "run_not_found"}
            return {
                "ok": True,
                "data": {
                    "summary": detail.summary.__dict__,
                    "artifacts": [artifact.__dict__ for artifact in detail.artifacts],
                },
            }
        if action == "update_provider":
            provider_id = str(data.get("provider_id", "")).strip()
            if not provider_id:
                return {"ok": False, "error": "provider_id_required"}
            env_values = data.get("env_values", {})
            if not isinstance(env_values, dict):
                return {"ok": False, "error": "env_values_must_be_object"}
            preferred = bool(data.get("preferred", False))
            status = self.service.update_provider_config(
                provider_id=provider_id,
                env_values={str(key): str(value) for key, value in env_values.items()},
                preferred=preferred,
            )
            if status is None:
                return {"ok": False, "error": "provider_not_found"}
            return {"ok": True, "data": status.to_dict()}
        if action == "install_provider":
            provider_id = str(data.get("provider_id", "")).strip()
            if not provider_id:
                return {"ok": False, "error": "provider_id_required"}
            try:
                result = self.service.install_provider_dependency(provider_id)
            except ValueError as error:
                return {"ok": False, "error": str(error)}
            return {"ok": True, "data": result.to_dict()}
        if action == "start_run":
            project_id = str(data.get("project_id", "")).strip()
            prompt = str(data.get("prompt", ""))
            try:
                result = self.service.start_run(project_id=project_id, prompt=prompt)
            except ValueError as error:
                return {"ok": False, "error": str(error)}
            return {"ok": True, "data": result.to_dict()}
        if action == "start_run_async":
            project_id = str(data.get("project_id", "")).strip()
            prompt = str(data.get("prompt", ""))
            try:
                job = self.service.start_run_async(project_id=project_id, prompt=prompt)
            except ValueError as error:
                return {"ok": False, "error": str(error)}
            return {"ok": True, "data": job.to_dict()}
        if action == "run_job_status":
            job_id = str(data.get("job_id", "")).strip()
            if not job_id:
                return {"ok": False, "error": "job_id_required"}
            job = self.service.get_run_job(job_id)
            if job is None:
                return {"ok": False, "error": "job_not_found"}
            return {"ok": True, "data": job.to_dict()}
        return {"ok": False, "error": f"unknown_action:{action}"}
